book/java-rebuilds-service.md

```diff
@@ -0,0 +1,154 @@
+
+Once it is enabled and the cache has started the JVM Build Service will watch for `PipelineRun` objects with the `JAVA_COMMUNITY_DEPENDENCIES` results. This will be present on end user builds that have dependencies from untrusted upstream sources.
+
+When one of these `PipelineRun` objects is observed JBS will extra the dependency list, and attempt to rebuild all the listed artifacts from source.
+
+The first stage of this process is to create an `ArtifactBuild` object. This object represents a Maven `group:artifact:version` (GAV) coordinate of the library that needs to be built from source.
+
+Once the `ArtifactBuild` object has been created JBS will then attempt to try and build it. This first step is to attempt to find the relevant source code.
+
+Once the source code location has been discovered a `DependencyBuild` object is created. There are generally less `DependencyBuild` objects than there are `ArtifactBuild`, as multiple artifacts can come from the same build. The controller will then try and build the artifact, first it will run a build discovery pipeline, that attempts to determine possible ways of building the artifact. Once discovery is complete it uses this information to attempt to build the artifact in a trial and error manner.
+
+If a build is successful the results are stored in a container image and the state is marked as complete, otherwise it is marked as failed and manual effort is required to fix the build.
+
+
+### Components
+
+JBS consists of the following components:
+
+**Controller**
+
+The controller is a Kubernetes controller written in go that orchestrates all aspects of the rebuild process. This will kick off all build related pipelines, and manage the state of the relevant Kubernetes objects.
+
+**Cache**
+
+The cache is a Quarkus Java application that caches artifacts from upstream repositories such as maven central. It also performs the following additional functions:
+
+- Handles container image based dependencies that have been rebuilt
+- Injects tracking metadata into class files to detect contaminated builds
+- Looks up source and build information from the build recipe database
+
+All dependency rebuilds are configured to only get their artifacts from the cache.
+
+**Build Request Processor**
+
+This is a multipurpose Quarkus CLI based app that performs quite a few different functions depending on the parameters it is invoked with. This is packaged up into an image, and then run in different parts of various pipelines by the operator.
+
+In essence any custom logic that is needed in a pipeline goes in here, and then the operator will invoke it with the correct arguments. This is a single multi-tool rather than individual functions as experience has shown this is a lot easier to manage. It's functions include:
+
+- Analysing a build and looking up its build information in the recipe database
+- Preprocessing a build to fix common problems, such as removing problematic plugins
```

Comment:
Are there examples of problematic plugins? Does this affect any functionality on the end application built? Are these problematic plugins customizable?

Reply:
One would be the nexus deploy plugin, which disables the standard deployment plugin and instead assumes you are deploying to a nexus server. They don't affect the functionality of the final application.